IEP-1000 Board doesn't match the target after cancel in debug/jtag tab

sigmaaa opened the pull request:
## Description

The initial bug can be reproduced this way:  select board in the debug/jtag tab -> click cancel -> select the previous target in the launch bar -> open the debug tab again -> board doesn't match the target. This happens because we had to save the target attribute of the configuration not only when we finish the dialog, but in the moment when the target is changed. We had to do this because we want the SVD path to always be according to the selected target. However, it led us to the problem when we couldn't actually add a custom SVD path, because it will be always reinitialized after the configuration is reopened. Moreover, it saves the wrong value for the SVD path, if the SVD path is not opened before the finish dialog.

To address this issue, I made the following changes: I removed the saving configuration in the listeners (which also ensures proper operation cancellation), and I introduced a dynamic variable for the SVD path. This variable will be replaced during the debug phase to resolve the problem

Fixes # ([IEP-1000](https://jira.espressif.com:8443/browse/IEP-1000))
Fixes #([IEP-1004](https://jira.espressif.com:8443/browse/IEP-1004))

## Type of change

Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How has this been tested?

Test 1:
- Edit debug configuration -> select different target -> click cancel ->  edit the same configuration again -> all pre-filled values should be correct
Test 2:
- Edit launch configuration -> select flash over jtag -> select different target ->  click finish
- Edit launch configuration again -> select different target -> click cancel -> edit the same configuration again -> all pre-filled 
values should be correct
Test 3:
- Create a new debug configuration
- in the SVD tab as path should be provided ${esp_svd_path} dynamic variable
- this variable should be present in the Variables... list
- debug the project with this configuration -> check peripherals view

**Test Configuration**:
* ESP-IDF Version:
* OS (Windows, Linux, and macOS):

## Dependent components impacted by this PR:

- SVD tab/ SVD path
- Debug configuration, Launch configuration

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [ ] Added Documentation
- [ ] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

New Features:
- Introduced a new system for resolving the System View Description (SVD) path, enhancing the software's ability to locate necessary files.
- Added a feature to revert target changes in the launch bar when the "Cancel" button is pressed, improving user control over target selection.
- Implemented a new method to handle different cases based on the active launch configuration and target, improving the software's adaptability.

Refactor:
- Simplified the logic in the `CMakeMainTab2` class by removing unnecessary code blocks and introducing more efficient methods.
- Streamlined the `TabDebugger` class with updates to UI elements, tooltips, and methods, enhancing the user interface's intuitiveness.

Note: The remaining changes do not have a direct impact on end-users and are categorized as "Refactor" or "Chores."
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sigmaaa commented on bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/SvdPathResolver.java:
```diff
@@ -0,0 +1,86 @@
+public class SvdPathResolver implements IDynamicVariableResolver
+{
+
+	public String resolveValue(IDynamicVariable variable, String argument) throws CoreException
```
mostly code from the updateSvd method

sigmaaa reviewed (commented):
self reviwed

alirana01 commented on bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/LaunchBarListener.java:
```diff
@@ -98,25 +100,29 @@ private void update(String newTarget)
-						final String jtag_device_id = activeConfig.getAttribute("org.eclipse.cdt.debug.gdbjtag.core.jtagDeviceId", ""); //$NON-NLS-1$ //$NON-NLS-2$
-						if ((activeConfig.getAttribute(IDFLaunchConstants.FLASH_OVER_JTAG, false) && !jtagIgnored)
+						final String jtag_device_id = activeConfig
+								.getAttribute("org.eclipse.cdt.debug.gdbjtag.core.jtagDeviceId", ""); //$NON-NLS-1$ //$NON-NLS-2$
```
Wouldn't it be better to move the constant strings to separate properties or interface?

sigmaaa replied:
fixed, by providing better way to check if selected configuration is of debug type

alirana01 commented on bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/LaunchBarListener.java:
```diff
@@ -98,25 +100,29 @@ private void update(String newTarget)
-							String targetForJtagFlash = activeConfig.getWorkingCopy().getAttribute(IDFLaunchConstants.TARGET_FOR_JTAG, ""); //$NON-NLS-1$
-							if (!newTarget.equals(targetForJtagFlash)) 
+							String targetForJtagFlash = activeConfig.getWorkingCopy()
+									.getAttribute(IDFLaunchConstants.TARGET_FOR_JTAG, ""); //$NON-NLS-1$
```
I think we are can use StringUtil.EMPTY string for blank or empty defaults can keep it clean and avoid unnecessary NLS tags

sigmaaa replied:
done

alirana01 reviewed (changes requested):
Nicely done and good utilization of Dynamic Variable Resolver, I didnt think of that when implementing just some minor comments from me to clean up a  bit.

sigmaaa commented:
Hi, @alirana01, thanks for the review. Fixed mentioned points in the latest commit

kolipakakondal commented on bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/SvdPathResolver.java:
```diff
@@ -0,0 +1,86 @@
+			String jarPath = new File(TabSvdTarget.class.getProtectionDomain().getCodeSource().getLocation().toURI())
+					.getPath();
+			if (!jarPath.contains(".jar")) //$NON-NLS-1$
+			{
+				selectedTargetPath = new File(FileLocator.resolve(svdUrl).toURI()).getPath();
+			}
+			else
+			{
+				IProject project = EclipseUtils
+						.getProjectByLaunchConfiguration(launchBarManager.getActiveLaunchConfiguration());
+				IFolder svdFolder = project.getFolder(IDFConstants.BUILD_FOLDER).getFolder("svd"); //$NON-NLS-1$
+				if (!svdFolder.exists())
+				{
+					svdFolder.create(true, true, new NullProgressMonitor());
+				}
+				IFile svdFile = project.getFolder(IDFConstants.BUILD_FOLDER).getFile(svdUrl.getPath());
+				if (!svdFile.exists())
+				{
+					JarFile jarFile = new JarFile(jarPath);
+					JarEntry file = (JarEntry) jarFile.getEntry(svdUrl.getFile().substring(1));
+					if (file != null)
+					{
+						InputStream inputStream = jarFile.getInputStream(file);
+						svdFile.create(inputStream, true, new NullProgressMonitor());
+						inputStream.close();
+					}
+					jarFile.close();
+				}
+				project.refreshLocal(IProject.DEPTH_INFINITE, new NullProgressMonitor());
+				selectedTargetPath = svdFile.getRawLocation().toOSString();
+			}
+		}
+		catch (Exception e)
+		{
+			Logger.log(e);
+		}
+		return selectedTargetPath;
+	}
+
+}
```
See if we can divide this code into two testable methods.

resolveSvdPath(String target)
resolveSvdPathFromJar(String target)

Something of this sort - 
```
public String resolveSvdPath(String target) throws Exception {
        URL svdUrl = Platform.getBundle(Activator.PLUGIN_ID)
                .getResource("svd/".concat(target.concat(".svd"))); //$NON-NLS-1$ //$NON-NLS-2$
        if (!svdUrl.toString().contains(".jar")) {
            return new File(FileLocator.resolve(svdUrl).toURI()).getPath();
        } else {
            return resolveSvdPathFromJar(target);
        }
    }
   ```

sigmaaa replied:
done

kolipakakondal commented on bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/SvdPathResolver.java:
```diff
@@ -0,0 +1,86 @@
+import com.espressif.idf.core.util.StringUtil;
+import com.espressif.idf.debug.gdbjtag.openocd.ui.TabSvdTarget;
+
+public class SvdPathResolver implements IDynamicVariableResolver
```
Add class comments

sigmaaa commented on bundles/com.espressif.idf.debug.gdbjtag.openocd/plugin.properties:
```diff
@@ -52,4 +52,5 @@ properties.mcu=ESP-IDF OpenOCD Path
-variable.description = Default C/C++ application (path to binaries) is determined automatically after build
\ No newline at end of file
+variable.description = Default C/C++ application (path to binaries) is determined automatically after build
+esp_svd_path_desc = The path to the SVD file associated with the selected target. Will be determined before joining the debug session
```
done

kolipakakondal reviewed (approved):
LGTM

sigmaaa commented:
> Hi @sigmaaa !
> 
> Tested under:
> 
> OS - Windows 10 ESP-IDF: master
> 
> Test 1 👍 Test 2 👍 Test 3 👍 Able to Debug/JTAG.
> 
> After pressing the "cancel" button in the debug configuration, there is no confusion between the target(ESP32) and board type. However, in the Launch Target, a switch occurs to a new (canceled target ESP32c3). As a result, if you try to build the project, you will encounter an error. There is nothing to worry about - you just need to manually change the target back to the one that was used before. Nevertheless, I believe this could potentially mislead the user.
> 
> I think we should consider how to automatically return to the original target.
> 
> ![image](https://user-images.githubusercontent.com/69584713/258162328-ac8ca964-82e9-47c7-9e50-81f09f07a0ef.png)

Hi @AndriiFilippov, @kolipakakondal , @alirana01 
added a workaround [fix: added workaround to roll back target after Cancel](https://github.com/espressif/idf-eclipse-plugin/pull/799/commits/c06d09aaaae5364b2fde466458825a819a1de502) for this issue with a `TODO` comment. Unfortunately, we don't have access to the Cancel button, which is part of LaunchConfigurationDialog and we don't have an extension point to this.

AndriiFilippov commented:
@sigmaaa hi ! 

Tested under: 
OS - Windows 10 
ESP-IDF: v5.1.1 


In general, everything is good, except for one scenario:

create new project -> go to Edit Launch Config -> click "Cancel" -> the Launch Mode and  Launch Target disappear. 

![image](https://github.com/espressif/idf-eclipse-plugin/assets/69584713/5997b72b-4cb2-4fb7-9178-8170259e4091)

sigmaaa commented:
Hi @AndriiFilippov, thanks for noticing. Fixed

AndriiFilippov commented:
Hi @sigmaaa ! 

please, check this scenario: 

create project -> go to "New Launch Configuration" ->  select new config: (Run + ESP-IDF Application) -> click "Next" -> click "Cancel" ->  the Launch Mode and Launch Target disappear.

AndriiFilippov commented:
@sigmaaa hi ! 

this issue is not related to your PR, my bad. 

LGTM 👍 



> Hi @sigmaaa !
> 
> please, check this scenario:
> 
> create project -> go to "New Launch Configuration" -> select new config: (Run + ESP-IDF Application) -> click "Next" -> click "Cancel" -> the Launch Mode and Launch Target disappear.
